### PR TITLE
feat(model): clear output when larger than MAX_STRING_LENGTH

### PR DESCRIPTION
**Pull request by fannheyward**
https://github.com/fannheyward/coc-rust-analyzer/issues/98

### CODE REVIEW

**Comment by chemzqm** [comment]
We should not just clear, but use latest lines  as much as possible.

File: src/model/outputChannel.ts [hunk 1]
```diff
@@ -41,8 +41,8 @@ export default class BufferChannel implements OutputChannel {
   }
 
   public append(value: string): void {
-    if (this._content.length > constants.MAX_STRING_LENGTH) {
-      this.clear()
+    if (this._content.length >= constants.MAX_STRING_LENGTH) {
```
Comment (chemzqm) [inline src/model/outputChannel.ts:44]: Should be `this._content.length + value.length`

File: src/model/outputChannel.ts [hunk 2]
```diff
@@ -51,20 +51,25 @@ export default class BufferChannel implements OutputChannel {
   }
 
   public appendLine(value: string): void {
-    if (this._content.length > constants.MAX_STRING_LENGTH) {
-      this.clear()
+    if (this._content.length >= constants.MAX_STRING_LENGTH) {
```
Comment (chemzqm) [inline src/model/outputChannel.ts:54]: Should be `this._content.length + value.length`

**Comment by miuirussia** [comment]
```
  src/model/outputChannel.ts:2:10 - error TS2305: Module '"buffer"' has no exported member 'constants'.
  
  2 import { constants } from 'buffer'
             ~~~~~~~~~
  
  
  Found 1 error.
```

**Comment by fannheyward** [comment]
@miuirussia what's your node version?

**Comment by miuirussia** [comment]
12.16.1

**Comment by fannheyward** [comment]
Can't reproduce. constants was added in 8.2.0 https://nodejs.org/api/buffer.html#buffer_buffer_constants, should not cause this error.

**Comment by chemzqm** [comment]
Remove node_modules and run `yarn install` in coc.nvim's folder.

**Comment by miuirussia** [comment]
Maybe in package.json need to declare nodejs version?

**Pull request merged**